### DISCUSSION TITLE
API for code search

### DISCUSSION THREAD

**Discussion by kendallm** [community/community#9991] [Other Feature Feedback, Questions, & Ideas]
Is there an API for interacting with Code Search?

**Comment by look** [comment]
@kendallm not yet, but we are working on it. Can you let us know some of the use cases you'd like to see? Thanks!

**Reply by shawnacscott** [reply]
One common need I have at work is to search all our repos  for a particular string. Generally, this is either for research purposes, or in order to programmatically generate pull requests to update something consistently across all our projects. Here are some examples:

* A new Ruby version is released. Since our policy is to only support a certain number of most-recent Ruby versions, this means that the oldest previously-supported Ruby version is now considered deprecated. I would like to be able to use search programmatically to generate a list of any applications that have are still running on the now-deprecated version and notify the owning teams of their timeline for upgrade. In addition, we automatically update Ruby versions on our internal gems, so I would like to search all of our gems `.ruby-version` files and create a PR to update them to the newly released version.
* I am working on modifying or deprecating an existing library method or API. I would like to search all of my organization's repos to generate a list of applications that are currently using the method/API that we are moving away from. 
* I would like to search all of our repos by the date they were last updated date to see if there are any stale repositories we may be able to delete or archive
* I would like to search all of our repos' lockfiles for various package managers by the date they were last updated date to ensure application dependencies are being updated with regularity
* I would like to be able to identify repos that have a specific file (for instance, `CODEOWNERS`) in a non-standard place so that those repos can be changed to be consistent
* If a new team is created or an old team renamed, I would like to combine my company's internal mapping of applications to teams with GitHub search to ensure that every repo's `CODEOWNERS` file declares a team that still exists so that it is always clear what team to ask for help if I need information about an application that y team does not own

One other thing I'll note is that you may notice that many of these use cases are not solely about searching. My ideal feature set would be some combination of search and Dependabot--more like a "find and replace" API, or some way to declare a specific file managed by an upstream resource. There are definitely other ways of accomplishing this that I use every day, but they are never quite as good as I would like them to be. 

Also, just want to say thank you for how nicely the search is working so far! It is already leaps and bounds improved, and works just like I always wanted the old version to work. I'm stoked at the features that already exist, and can't wait to see more!

**Reply by kendallm** [reply]
I am currently building out a tagging system for some internal repos. I plan on creating a dashboard across all of our repos that aggregates tags and links them together. This is sort of possible with the current search endpoint, but it is not straight forward to get to exact lines of code with our current call pattern using the current search API.

Another use case is finding usage of library versions across repos in an org. Ideally this could be scoped to files that declare locked dependencies across repos, but I would also want to look at transitive dependencies brought in by direct dependencies.

**Reply by mario-guerra** [reply]
I would like to +1 the request for an API. We are deprecating a large number of old packages and we'd like to be able to programmatically find repositories that are making use of the deprecated packages so we can notify the repo owners that they should upgrade.

**Reply by sgandon** [reply]
One main use case that we have and similar to @kendallm is that we want to upgrade a library or compile it with a more recent SDK version or deprecate a given library and we would like to be able to find all the repositories that would reference this library in order to contact the owner and make sur upgrading is possible.
and I believe an API would allow us to get response payload that we can furthermore exploit for filtering like only give us the distinct repositories for this search result.

**Reply by jetersen** [reply]
There are simply some searches that are impossible in the old elastic search code search. I really wish this had an API. I may have to default to HTML scraping of cs.github.com 😂

**Reply by gabrielenosso** [reply]
I can share my use case: I am working on a Software Catalog to increase the Reusability Index in my Company.
Using the Code Search API would give us a chance to find them easily with the search of specific strings (For instance, in the package.json files of repos).
Can't do that with the "normal" API, as special characters are not supported.

**Reply by jpivarski** [reply]
I'm a developer of software tools for physicists and, as a part of that, I need quantitative metrics of software adoption and patterns of use among those physicists. I've used a [variety of tools to gather this sort of data](https://indico.cern.ch/event/1138009/) (PyPI/conda-forge download statistics, regex searches of our conference talks, GitHub stars, surveys), the [most incisive metrics are derived from the GitHub API](https://indico.fnal.gov/event/43829/contributions/193602/).

Search results are awesome: the most relevant metric is "What code are the physicists actually writing?" (Assuming they put all their code on GitHub, which is close to true nowadays.) But the GitHub Search API is pretty limited and I've needed to do queries in a hacky way... and slowly pull data over the course of hours to avoid rate limits (and secondary rate limits). In part, that was limited by old GitHub search semantics.

I've used the new code search enough to know that I'll be able to write more directed searches with this. I'll want to get the results as text-based lists, and _could_ scrape the HTML, but it would be much nicer to use a formal API.

While we're at it, I usually don't need the full set of information that the GitHub API returns for its results—a page full of JSON for each repo—most of the results that I'm looking for are lists of repos (JSON array of strings) and lists of users (JSON array of strings).

Unlike other use-cases, I'm interested in doing a one-time pull, maybe followed up several months later, not any kind of continuous pipeline.

**Comment by OnkarRuikar** [comment]
Many of the repos/projects have their dedicated websites. The CS API will allow them to provide more customized code search into the repo directly on their websites.

***

With the API any website could embed the Code Search. They could have more user friendly querying mechanism. The results could be shown in more creative ways.

*** 

Just a thought, with a clever repo structure and queries, using the API, the content on GitHub could be used as a read only database.

**Comment by gabrielenosso** [comment]
I can share my use case: I am working on a Software Catalog to increase the Reusability Index in my Company.
Using the Code Search API would give us a chance to find them easily with the search of specific strings (For instance, in the package.json files of repos).
Can't do that with the "normal" API, as special characters are not supported.

**Reply by ng-druid** [reply]
Integrating [JSON path](https://learn.microsoft.com/en-us/sql/relational-databases/json/json-path-expressions-sql-server?view=sql-server-ver16) into the search API would enable package.json searching and much more. Map reduce support using custom JavaScript functions passed as JSON encoded strings would also be welcomed.

**Comment by lachilles** [comment]
Hi @look , is the code search API available yet? I have access to the code search beta and am able to view all the search results I want from the web UI. Would love to have this available programmatically since we have similar use cases to others. Would like to search spring-app-chart versions org wide and find outdated versions. If there's an open issue somewhere that I can subscribe to and be notified when the feature is ready, please share it. Thank you!!

**Reply by look** [reply]
Hi @lachilles, not yet, we are still working on scaling the system up before we can tackle the API. I will update this issue when we have news to share. Sorry, but thank you for your patience!
